File: backends/fastapi/app/usecase/queue.py
from __future__ import annotations

from datetime import timedelta

from app.domain import errors
from app.domain.models import QueueStatus, QueueToken
from app.usecase.ports import (
    Clock,
    EventRepository,
    IDGenerator,
    QueueRepository,
    RateLimiter,
)


class QueueService:
    """The pressure valve. Only an admitted token may proceed to checkout; that gate
    is enforced by ReservationService via is_admitted.
    """

    def __init__(
        self,
        queue: QueueRepository,
        events: EventRepository,
        limiter: RateLimiter,
        clock: Clock,
        ids: IDGenerator,
        admit_batch: int,
    ):
        self._queue = queue
        self._events = events
        self._limiter = limiter
        self._clock = clock
        self._ids = ids
        self._admit_batch = admit_batch if admit_batch > 0 else 50

    async def join(self, user_id: str, event_id: str) -> QueueToken:
        if await self._events.find_by_id(event_id) is None:
            raise errors.NOT_FOUND

        # Rate limit joins so a script cannot spam the queue endpoint.
        if not await self._limiter.allow(f"queue_join:{user_id}:{event_id}", 5, timedelta(minutes=1)):
            raise errors.RATE_LIMITED

        existing = await self._queue.find(user_id, event_id)
        if existing is not None:
            return await self._decorate(existing)

        pos = await self._queue.next_position(event_id)
        token = QueueToken(
            id=self._ids.new_id(),
            user_id=user_id,
            event_id=event_id,
            position=pos,
            status=QueueStatus.WAITING,
        )
        await self._queue.upsert(token)
        return await self._decorate(token)

    async def status(self, user_id: str, event_id: str) -> QueueToken:
        token = await self._queue.find(user_id, event_id)
        if token is None:
            raise errors.NOT_FOUND
        return await self._decorate(token)

    async def is_admitted(self, user_id: str, event_id: str) -> bool:
        token = await self._queue.find(user_id, event_id)
        if token is None:
            return False
        token = await self._decorate(token)
        return token.status == QueueStatus.ADMITTED

    async def _decorate(self, token: QueueToken) -> QueueToken:
        # Flip waiting -> admitted once position is within the batch, and persist it.
        if token.status == QueueStatus.WAITING and token.position < self._admit_batch:
            token.status = QueueStatus.ADMITTED
            token.admitted_at = self._clock.now()
            await self._queue.upsert(token)
        return token

File: backends/go/internal/usecase/queue.go
package usecase

import (
	"context"
	"time"

	"github.com/ticketing-labs/backend-go/internal/domain"
)

// QueueService is the pressure valve. Under a flash sale you cannot let everyone hit
// the reservation path at once, so they take a numbered ticket and wait. Only an
// admitted token may proceed to checkout; that gate is enforced by ReservationService.
type QueueService struct {
	queue   QueueRepository
	events  EventRepository
	limiter RateLimiter
	clock   Clock
	ids     IDGenerator
	// admitUpTo is the position below which tokens are considered admitted. In a real
	// system a background process raises this as capacity frees up; here it is a simple
	// batch size so the demo is deterministic.
	admitBatch int
}

func NewQueueService(q QueueRepository, e EventRepository, rl RateLimiter, c Clock, ids IDGenerator, admitBatch int) *QueueService {
	if admitBatch <= 0 {
		admitBatch = 50
	}
	return &QueueService{queue: q, events: e, limiter: rl, clock: c, ids: ids, admitBatch: admitBatch}
}

func (s *QueueService) Join(ctx context.Context, userID, eventID string) (*domain.QueueToken, error) {
	if _, err := s.events.FindByID(ctx, eventID); err != nil {
		return nil, domain.ErrNotFound
	}

	// Rate limit joins per user+event so a script cannot spam the queue endpoint.
	allowed, err := s.limiter.Allow(ctx, "queue_join:"+userID+":"+eventID, 5, time.Minute)
	if err == nil && !allowed {
		return nil, domain.ErrRateLimited
	}

	// Idempotent-ish: re-joining returns the existing token rather than a new place.
	if existing, err := s.queue.Find(ctx, userID, eventID); err == nil && existing != nil {
		return s.decorate(existing), nil
	}

	pos, err := s.queue.NextPosition(ctx, eventID)
	if err != nil {
		return nil, domain.ErrInternal
	}
	token := &domain.QueueToken{
		ID:       s.ids.NewID(),
		UserID:   userID,
		EventID:  eventID,
		Position: pos,
		Status:   domain.QueueWaiting,
	}
	if err := s.queue.Upsert(ctx, token); err != nil {
		return nil, domain.ErrInternal
	}
	return s.decorate(token), nil
}

func (s *QueueService) Status(ctx context.Context, userID, eventID string) (*domain.QueueToken, error) {
	token, err := s.queue.Find(ctx, userID, eventID)
	if err != nil || token == nil {
		return nil, domain.ErrNotFound
	}
	return s.decorate(token), nil
}

// decorate flips a waiting token to admitted once its position is within the batch,
// stamping admitted_at. This keeps admission logic in one place.
func (s *QueueService) decorate(t *domain.QueueToken) *domain.QueueToken {
	if t.Status == domain.QueueWaiting && t.Position < s.admitBatch {
		now := s.clock.Now()
		t.Status = domain.QueueAdmitted
		t.AdmittedAt = &now
		_ = s.queue.Upsert(context.Background(), t)
	}
	return t
}

// IsAdmitted is used by ReservationService to enforce the checkout gate.
func (s *QueueService) IsAdmitted(ctx context.Context, userID, eventID string) bool {
	t, err := s.queue.Find(ctx, userID, eventID)
	if err != nil || t == nil {
		return false
	}
	t = s.decorate(t)
	return t.Status == domain.QueueAdmitted
}
